roxie/ccd/ccdfile.cpp

```diff
@@ -1270,6 +1621,211 @@ class CRoxieFileCache : implements IRoxieFileCache, implements ICopyFileProgress
+            E->Release();
+        }
+    }
+
+
+    virtual void warmOsCache(const char *cacheInfo) override
+    {
+        if (!cacheInfo)
+            return;
+        size_t os_page_size = getpagesize();
+        char t = 0;
+        unsigned touched = 0;
+        unsigned preloaded = 0;
+        Owned<const ITopologyServer> topology = getTopology();
+        while (*cacheInfo)
+        {
+            // MORE - what if there is a | in the filename?
+            unsigned fileChannel = strtoul(cacheInfo, (char **) &cacheInfo, 10);
+            if (*cacheInfo != '|')
+                break;
+            if (!topology->implementsChannel(fileChannel))
+            {
+                const char *eol = strchr(cacheInfo, '\n');
+                if (!eol)
+                    break;
+                cacheInfo = eol+1;
+                continue;
+            }
+            cacheInfo++;
+            const char *endName = strchr(cacheInfo, '|');
+            assert(endName);
+            if (!endName)
+                break;
+            StringBuffer fileName(endName-cacheInfo, cacheInfo);
+            Owned<IKeyIndex> keyIndex;
+            bool keyFailed = false;
+            unsigned fileIdx = (unsigned) -1;
+            char *file_mmap = nullptr;
+            int fd = open(fileName, 0);
+            struct stat file_stat;
+            if (fd != -1)
+            {
+                fstat(fd, &file_stat);
+                file_mmap = (char *) mmap((void *)0, file_stat.st_size, PROT_READ, MAP_SHARED, fd, 0);
+                if (file_mmap == MAP_FAILED)
+                {
+                    DBGLOG("Failed to map file %s to pre-warm cache (error %d)", fileName.str(), errno);
+                    file_mmap = nullptr;
+                }
+            }
+            else if (traceLevel)
+            {
+                DBGLOG("Failed to open file %s to pre-warm cache (error %d)", fileName.str(), errno);
+            }
+            Owned<ILazyFileIO> localFile = lookupLocalFile(fileName);
+            if (localFile)
+            {
+                fileIdx = localFile->getFileIdx();
+            }
+            cacheInfo = endName+1;  // Skip the |
+            while (*cacheInfo==' ')
+                cacheInfo++;
+            for (;;)
+            {
+                bool inNodeCache = (*cacheInfo=='*');
+                if (inNodeCache)
+                    cacheInfo++;
+                offset_t startPage = readPage(cacheInfo);
+                offset_t endPage;
+                if (*cacheInfo=='-')
+                {
+                    cacheInfo++;
+                    endPage = readPage(cacheInfo);
+                }
+                else
+                    endPage = startPage;
+                if (traceLevel > 4)
+                    DBGLOG("Touching %s %" I64F "x-%" I64F "x", fileName.str(), startPage, endPage);
+                startPage = startPage << CacheInfoEntry::pageBits;
+                endPage = (endPage+1) << CacheInfoEntry::pageBits;
```

Comment:
think it would clearer if used new variables, e.g. startOffset, endOffset , as being converted from page numbers to offsets.

Reply:
ok